text/0000-c-string-construction.md

```diff
@@ -0,0 +1,72 @@
+- Feature Name: c_string_from_iter
+- Start Date: 2015-02-26
+- RFC PR:
+- Rust Issue:
+
+# Summary
+
+Amend the methods available to construct a `CString` to improve composability
+and follow the conventions emerging elsewhere in the standard library.
+
+# Motivation
+
+The implementation of RFCs [#592][rfc 592] and [#840][rfc 840] has resolved
+most of the issues with the design of `std::ffi::CString`, but some new APIs
+have been created at the same time, needing stabilization. This proposal
+aims at addressing the following issues:
+
+1. The `IntoBytes` trait does not seem wholly justified: it falls short of
+   supporting `IntoIterator`, and has become yet another special-interest trait
+   to care about when designing APIs producing string-like data. 
+
+2. The exposure of `Vec` as an intermediate type for all conversions
+   precludes small-string optimizations, such as an in-place variant
+   implemented in [c_string](https://github.com/mzabaluev/rust-c-str).
+
+3. Stylistic: `Result` as return value type of `new` feels a bit too 'loaded'.
+   It's used in some other places, but the general expectation on `new` is to
+   be the most straightforward way to obtain a value of the type, while more
+   involved failure modes tend to be more typical on `from_*` constructors
+   and the like.
+
+[rfc 592]: https://github.com/rust-lang/rfcs/pull/592
+[rfc 840]: https://github.com/rust-lang/rfcs/pull/840
+
+# Detailed design
+
+Replace the constructor accepting `IntoBytes` with one accepting
+`IntoIterator`, following the `from_iter` pattern in collection types:
+
+```rust
+impl CString {
+    pub fn from_iter<I>(iterable: I) -> Result<CString, NulError>
+        where I: IntoIterator<Item=u8>
```

Comment:
I feel like `from_iter` may want a different name as these look a bit odd to me:

```rust
CString::from_iter(b"foo");
CString::from_iter("foo");
CString::from_iter(my_string);
```

(none of them actually have to do with iterators)

Also, with an `IntoIterator<Item=u8>` bound, I think that this may break `CString::new("foo")` (e.g. construction from a `str`) because implementing `IntoIterator<Item=u8>` for `str` probably won't get too far. For example it would enable code like:

```rust
for byte in str {
    println!("{}", byte);
}
```

It's quite debatable whether `byte` should indeed be a byte or perhaps a `char` (hence the requirement that it must be chosen).